src/sema/eval.rs

```diff
@@ -196,3 +201,160 @@ pub fn eval_const_rational(
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                overflow_check(ns, left.sub(right), ty.clone(), *loc);
+                bigint_to_expression(loc, ty, left.sub(right))
+            } else {
+                Expression::Subtract(
+                    *loc,
+                    ty.clone(),
+                    *unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                )
+            }
+        }
+
+        Expression::Multiply(loc, ty, unchecked, left, right) => {
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                overflow_check(ns, left.mul(right.to_u32().unwrap()), ty.clone(), *loc);
+                bigint_to_expression(loc, ty, left.mul(right.to_u32().unwrap()))
+            } else {
+                Expression::Multiply(
+                    *loc,
+                    ty.clone(),
+                    *unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                )
+            }
+        }
+
+        Expression::Power(loc, ty, unchecked, left, right) => {
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                overflow_check(ns, left.pow(right.to_u32().unwrap()), ty.clone(), *loc);
+                bigint_to_expression(loc, ty, left.pow(right.to_u32().unwrap()))
+            } else {
+                Expression::Power(
+                    *loc,
+                    ty.clone(),
+                    *unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                )
+            }
+        }
+
+        Expression::ShiftLeft(loc, ty, left, right) => {
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                overflow_check(ns, left.shl(right.to_u32().unwrap()), ty.clone(), *loc);
+                bigint_to_expression(loc, ty, left.shl(right.to_u32().unwrap()))
+            } else {
+                Expression::ShiftLeft(*loc, ty.clone(), Box::new(left), Box::new(right))
+            }
+        }
+        _ => expr.clone(),
```

Comment:
You should also be evaluating Modulo, ShiftRight, Trunc, SignExt, ZeroExt, Cast. These can also take part in arithmetic expressions and can be evaluated during compile time.

@seanyoung Should Bitwise and Boolean operations be here too?

Reply:
Since the result is thrown away, I'm only evaluating operations that can result in overflow. I think evaluating the mentioned expressions will not be needed here

Reply:
> Since the result is thrown away, I'm only evaluating operations that can result in overflow. I think evaluating the mentioned expressions will not be needed here

`int8 c = 255 + (2 >> 1)`. This has shift right and is overflowing. You can find examples of this for all the expressions I mentioned.

Reply:
Devide is missing too. Just because an expression can be classified as a decrescent function, it does not meant the overall expression it belongs to won't overflow.